sabnzbd/decoder.py

```diff
@@ -154,41 +162,50 @@ def run(self):
                 sabnzbd.NzbQueue.reset_try_lists(article)
                 continue
 
-            except CrcError as crc_error:
-                logging.info("CRC Error in %s" % art_id)
-
+            except BadData as error:
                 # Continue to the next one if we found new server
                 if search_new_server(article):
                     continue
 
                 # Store data, maybe par2 can still fix it
-                decoded_data = crc_error.data
+                decoded_data = error.data
+
+            except BadUu:
+                logging.info("Badly formed uu article in %s", art_id)
+
+                # Try the next server
+                if search_new_server(article):
+                    continue
 
             except (BadYenc, ValueError):
                 # Handles precheck and badly formed articles
                 if nzo.precheck and raw_data and raw_data[0].startswith(b"223 "):
                     # STAT was used, so we only get a status code
                     article_success = True
                 else:
-                    # Examine headers (for precheck) or body (for download)
-                    # Look for DMCA clues (while skipping "X-" headers)
-                    # Detect potential UUencode
-                    for line in raw_data:
-                        lline = line.lower()
-                        if b"message-id:" in lline:
+                    # Try uu-decoding
+                    if (not nzo.precheck) and article.nzf.type != "yenc":
+                        try:
+                            decoded_data = decode_uu(article, raw_data)
```

Comment:
It feels a bit wrong to have this here.
Shouldn't we set the `article.nzf.type` and then `sabnzbd.Decoder.process()`?

Reply:
I don't think so: we cannot be certain an article is uu, just because we get to this point. The yenc decoding attempt could have failed for some other reason, e.g. corrupt data or even a complete lack of an encoded payload (like in one of the examples you forwarded). Note that setting the nzf type to uu here - without having successfully decoded the article as such - would causes any remaining articles in the same nzf to never be tried with yenc.

Reply:
I actually meant to keep the old detection, so if we find `\nbegin` that would be the trigger.

Reply:
That approach only really works for singlepart uu, where every article comes with a begin line. With multipart, the entire nzf would be written off if the lowest_partnum article is missing/damaged/slow/just plain weird (e.g. that aislin nzb) even if all other articles were fine.

You did just give me ideas for some nasty functional tests though. 😀

Reply:
Well, in that case I would propose to try yenc always first. Even if we know the file is uu.
Yes it's a performance impact, but who cares in case of uu?

Reply:
> Even if we know the file is uu.
 
Why?? Continuing to try the other encoding after the `nzf.type` was successfully determined is pointless, unless uploaders were to start mixing encodings **within a single nzf**.

> Yes it's a performance impact, but who cares in case of uu?

I can't think of any user that wants to see a program intentionally waste resources for no gain whatsoever, as a punishment for downloading an nzb that just happens to use a less common encoding type. An encoding type they have no way to know in advance no less.

And I sure cared when writing the code: by avoiding futile decoding attempts (for either decoding type; notice how the fallback to uu is skipped once the type is set to yenc), and preventing the addition of uu support from having a measurable performance impact on a typical yenc-encoded job.